crates/fluvio-protocol-derive/src/ast/prop.rs

```
@@ -151,76 +143,324 @@ impl UnnamedProp {
+/// ````
+/// To set a specific type you can do this:
+/// ````ignore
+/// let ident_type = Ident::new("u8", Span::call_site());
+/// let func_value = prop_attrs_type_value(prop_attr_type, Some(&ident_type))
+/// ````
+///
+pub fn prop_attrs_type_value(
+    attrs_type: Option<&PropAttrsType>,
+    ident_type: Option<&Ident>,
+) -> TokenStream {
+    if let Some(attr) = attrs_type {
+        match &attr {
+            PropAttrsType::Lit(data) => parse_quote!(#data),
+            PropAttrsType::Fn(data) => TokenStream::from_str(&format!("{}()", data)).unwrap(),
+            PropAttrsType::Int(data) => {
+                if let Some(itype) = ident_type {
+                    TokenStream::from_str(&format!("{}_{}", data, itype)).unwrap()
+                } else {
+                    // By default it's i16, because most places use it
+                    TokenStream::from_str(&format!("{}_i16", data)).unwrap()
+                }
+            }
         }
-        (None, None) if min < 0 => Some(format!("Min version({min}) must be positive.")),
-        _ => None,
+    } else {
+        parse_quote!(0_i16)
+    }
+}
+/// A type that will handle the values passed in properties
+/// and convert them later on to TokenStream.
+///
+/// Using this type allows you to pass values multiple ways:
+/// # Example
+///
+/// ```ignore
+/// // Constant as a path
+/// const TEST: i16 = 1;
+/// #[fluvio(min_version = TEST)]
+/// ```
+///
+/// ```ignore
+/// // Constant as a literal
+/// const TEST: i16 = 1;
+/// #[fluvio(min_version = "TEST")]
+/// ```
+///
+/// ```ignore
+/// // Function as a literal
+/// fn test() -> i16 { 1 }
+/// #[fluvio(min_version = "test()")]
+/// ```
+///
+/// ```ignore
+/// // Int
+/// #[fluvio(min_version = 1)]
+/// ```
+///
+/// None has a default Int value of 0 which is set in prop_attrs_type_value
+#[derive(Clone)]
+pub enum PropAttrsType {
+    Lit(Ident),
+    Fn(Ident),
+    Int(i16),
+}
+impl Default for PropAttrsType {
```

Why do we need to impl `Default` for attr type? If the type is not identifiable, we should just throw an error.

---

I was just following how it worked previously that it defaults to 0 so decided to set a default value as well.

---

If you mean that previously default value of `min_version: i16` is 0, that's fine. The equivalent of that for new `min_version: Option<PropAttrsType>` is None. This doesn't mean that the default needs for `PropAttrsType ` type.

---

I removed the default with the new changes.